sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/batch/implementation/BatchAsyncContainerExecutor.java

```diff
@@ -0,0 +1,275 @@
+        final int maxOperationCount,
+        final int maxPayloadLength) {
+
+        checkNotNull(container, "expected non-null containerCore");
+        checkArgument(maxOperationCount > 0, "expected maxServerRequestOperationCount > 0, not %s", maxOperationCount);
+        checkArgument(maxPayloadLength > 0, "expected maxServerRequestBodyLength > 0, not %s", maxPayloadLength);
+
+        this.container = container;
+        this.maxOperationCount = maxOperationCount;
+        this.maxPayloadLength = maxPayloadLength;
+        this.docClientWrapper = CosmosBridgeInternal.getAsyncDocumentClient(container.getDatabase());
+        this.throttlingRetryOptions = docClientWrapper.getConnectionPolicy().getThrottlingRetryOptions();
+    }
+
+    public Mono<TransactionalBatchOperationResult<?>> addAsync(
+        final ItemBatchOperation<?> operation,
+        final RequestOptions options) {
+
+        checkNotNull(operation, "expected non-null operation");
+
+        CompletableFuture<TransactionalBatchOperationResult<?>> future = this.validateAndMaterializeOperation(operation, options)
+            .thenComposeAsync(
+                t -> this.resolvePartitionKeyRangeIdAsync(operation)
+                    .toFuture().thenCompose(
+                        (String resolvedPartitionKeyRangeId) -> {
+
+                            BatchAsyncStreamer streamer = this.getOrAddStreamerForPartitionKeyRange(resolvedPartitionKeyRangeId);
+
+                            ItemBatchOperationContext context = new ItemBatchOperationContext(
+                                resolvedPartitionKeyRangeId,
+                                BatchAsyncContainerExecutor.getRetryPolicy(this.throttlingRetryOptions));
+                            operation.attachContext(context);
+
+                            streamer.add(operation);
+
+                            return context.getOperationResultFuture();
+                        }));
+
+        return Mono.fromFuture(future);
+    }
+
+    public CompletableFuture<Boolean> validateAndMaterializeOperation(
+        final ItemBatchOperation<?> operation,
+        final RequestOptions options) {
+
+        if (options != null) {
+            checkState(options.getConsistencyLevel() == null
+                && options.getPostTriggerInclude() == null
+                && options.getPreTriggerInclude() == null
+                && options.getSessionToken() == null, "UnsupportedBulkRequestOptions");
+
+            BatchAsyncContainerExecutor.validateOperationEpk(operation, options);
+        }
+
+        return operation.materializeResource();
+    }
+
+    private Mono<PartitionKeyRangeBatchExecutionResult> executeAsync(PartitionKeyRangeServerBatchRequest request) throws Exception {
+        request.setAtomicBatch(false);
+        request.setShouldContinueOnError(true);
+
+        final Semaphore limiter = this.getOrAddLimiterForPartitionKeyRange(request.getPartitionKeyRangeId());
+        limiter.acquire();
```

Comment:
using a semaphore in a async method will result the thread to block.

Have we consider the fact that this can run on an io thread? and that can cause problem?

Reply:
We do have timeout for requests right? So it will release eventually. And we run it in a background. Still this will be an issue?